### PR TITLE
Use ValueStringBuilder for forwarded transform

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/microsoft/reverse-proxy/issues/799

### CODE REVIEW

File: src/ReverseProxy/Utilities/ValueStringBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Yarp.ReverseProxy.Utilities
+{
+    internal ref partial struct ValueStringBuilder
```
Comment (Tratcher) [inline src/ReverseProxy/Utilities/ValueStringBuilder.cs:8]: Add a comment with a permanent link to the original.

File: src/ReverseProxy/Utilities/ValueStringBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
+using System;
```
Comment (Tratcher) [inline src/ReverseProxy/Utilities/ValueStringBuilder.cs:1]: License header.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestHeaderForwardedTransform.cs [hunk 3]
```diff
@@ -178,12 +178,12 @@ private void AppendNode(IPAddress ipAddress, int port, NodeFormat format, String
             if (addPort)
             {
                 builder.Append(':');
-                builder.Append(port);
+                builder.Append(port.ToString());
```
Comment (Tratcher) [inline src/ReverseProxy/Service/RuntimeModel/Transforms/RequestHeaderForwardedTransform.cs:181]: InvariantCulture

Comment (davidfowl) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestHeaderForwardedTransform.cs:181]: This allocation seems unnecessary.

Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestHeaderForwardedTransform.cs:181]: True, but ValueStringBuilder doesn't support ints.

It looks like StringBuilder calls into Number.Formatting:
https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs#L830

Comment (MihaZupan) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestHeaderForwardedTransform.cs]: In Uri code we do
```c#
const int MaxUshortLength = 5;
bool success = port.TryFormat(dest.AppendSpan(MaxUshortLength), out int charsWritten);
Debug.Assert(success);
dest.Length -= MaxUshortLength - charsWritten;
```

File: src/ReverseProxy/Utilities/ValueStringBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                RawChars[pos] = s[0];
+                _pos = pos + 1;
+            }
+            else
+            {
+                AppendSlow(s);
+            }
+        }
+
+        private void AppendSlow(string s)
```
Comment (MihaZupan) [inline src/ReverseProxy/Utilities/ValueStringBuilder.cs:73]:
```suggestion
```

This part can be removed as the `Length == 1` case really _isn't_ common. We should fix this in the runtime too.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**